qiskit_experiments/test/mock_iq_backend.py

```diff
@@ -26,6 +26,147 @@
+        return Options(
+            shots=1024,
+            meas_level=MeasLevel.CLASSIFIED,
+            meas_return="single",
+        )
+
+    @staticmethod
+    def _get_state_strings(n_qubits: int) -> List[str]:
+        """Generate all state strings for the system."""
+        states, format_str = [], "{0:0" + str(n_qubits) + "b}"
+
+        for state_num in range(2**n_qubits):
+            states.append(format_str.format(state_num))
+
+        return states
+
+    @abstractmethod
+    def _compute_outcome_probabilities(self, circuits: List[QuantumCircuit]):
+        """Compute the probabilities of measuring 0 or 1 for each of the given
+         circuits based on the previous measurement shot.
+
+        This methods computes the dictionary self._precomputed_probabilities where
+        the keys are a tuple consisting of the circuit index and the previous outcome,
+        e.g. "0" or "1" for a single qubit. The values are the corresponding probabilities.
+
+        Args:
+            circuits: The circuits from which to compute the probabilities.
+        """
+
+    def run(self, run_input, **options):
+        """Run the restless backend."""
+
+        self.options.update_options(**options)
+        shots = self.options.get("shots")
+        meas_level = self.options.get("meas_level")
+
+        result = {
+            "backend_name": f"{self.__class__.__name__}",
+            "backend_version": "0",
+            "qobj_id": 0,
+            "job_id": 0,
+            "success": True,
+            "results": [],
+        }
+
+        self._compute_outcome_probabilities(run_input)
+
+        num_qubits = run_input[0].num_qubits
+
+        prev_outcome = "0" * num_qubits
+        state_strings = self._get_state_strings(num_qubits)
+
+        # Setup the list of dicts where each dict corresponds to a circuit.
+        sorted_memory = [{"memory": [], "metadata": circ.metadata} for circ in run_input]
+
+        for _ in range(shots):
+            for circ_idx, _ in enumerate(run_input):
+                prob = self._precomputed_probabilities[(circ_idx, prev_outcome)]
+                # Generate the next shot dependent on the pre-computed probabilities.
+                outcome = self._rng.choice(state_strings[:2], p=[1 - prob, prob])
+                # Append the single shot to the memory of the corresponding circuit.
+                sorted_memory[circ_idx]["memory"].append(hex(int(outcome, 2)))
+
+                prev_outcome = outcome
+
+        for idx, circ in enumerate(run_input):
+            ones = sorted_memory[idx]["memory"].count("0x1")
+            run_result = {
+                "shots": shots,
+                "success": True,
+                "header": {"metadata": circ.metadata},
+                "meas_level": meas_level,
+                "data": {
+                    "counts": {"1": ones, "0": shots - ones},
+                    "memory": sorted_memory[idx]["memory"],
+                },
+            }
```

Comment:
I think we need the counts, since we want to test a restless experiment with a standard processor:

```python
standard_processor = DataProcessor("counts", [Probability("1")])
amp_exp.analysis.set_options(data_processor=standard_processor)
```

We probably need to generalise this to more than one qubit.

Reply:
I adapted the code here: [55a8183](https://github.com/Qiskit/qiskit-experiments/pull/689/commits/55a818331f55628f71e9ea2cfcc709025d6734f4).